File: front/lib/domain/model/producer_account.dart
import 'package:amap_en_ligne/domain/model/organization.dart';
import 'package:amap_en_ligne/domain/model/product_type.dart';
import 'package:amap_en_ligne/domain/model/user_preferences.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'producer_account.freezed.dart';
part 'producer_account.g.dart';

enum ProducerManagementMode {
  @JsonValue('ACCOUNT_BACKED')
  accountBacked,
  @JsonValue('NO_ACCOUNT')
  noAccount,
}

/// Producer-side embedding: one entry per `Organization` the producer is
/// linked to. Mirrors the back `ProducerOrganization` data class (note: the
/// reverse `OrganizationProducer` in `organization.dart` has `producer_account_id`
/// — these are intentionally distinct wire shapes).
@freezed
abstract class ProducerOrganization with _$ProducerOrganization {
  const factory ProducerOrganization({
    @JsonKey(name: 'organization_id') required String organizationId,
    // ISO-8601 instant string, e.g. "2026-05-18T22:23:25.095Z".
    @JsonKey(name: 'association_instant') required String associationInstant,
    required OrganizationProducerStatus status,
  }) = _ProducerOrganization;

  factory ProducerOrganization.fromJson(Map<String, Object?> json) =>
      _$ProducerOrganizationFromJson(json);
}

@freezed
abstract class ProducerProduct with _$ProducerProduct {
  const factory ProducerProduct({
    required String name,
    @JsonKey(name: 'product_type_id') required String productTypeId,
    @JsonKey(name: 'supported_basket_sizes')
    @Default([])
    List<BasketSize> supportedBasketSizes,
    String? description,
  }) = _ProducerProduct;

  factory ProducerProduct.fromJson(Map<String, Object?> json) =>
      _$ProducerProductFromJson(json);
}

@freezed
abstract class ProducerAccount with _$ProducerAccount {
  const factory ProducerAccount({
    @JsonKey(name: 'producer_account_id') required String producerAccountId,
    required String name,
    @JsonKey(name: 'contact_email') String? contactEmail,
    String? address,
    String? website,
    @JsonKey(name: 'active_status') @Default(true) bool activeStatus,
    // ISO-8601 instant strings (e.g. "2026-05-18T22:23:25.093Z") — matches
    // the back's kotlin.time.Instant default serialization.
    @JsonKey(name: 'created_instant') String? createdInstant,
    @JsonKey(name: 'last_updated_instant') String? lastUpdatedInstant,
    @JsonKey(name: 'management_mode')
    @Default(ProducerManagementMode.accountBacked)
    ProducerManagementMode managementMode,
    @JsonKey(name: 'linked_producer_account')
    LinkedProducerAccount? linkedProducerAccount,
    @Default([]) List<ProducerProduct> products,
    @Default([]) List<ProducerOrganization> organizations,
    @JsonKey(name: 'user_preferences') UserPreferences? userPreferences,
  }) = _ProducerAccount;

  factory ProducerAccount.fromJson(Map<String, Object?> json) =>
      _$ProducerAccountFromJson(json);
}

@freezed
abstract class LinkedProducerAccount with _$LinkedProducerAccount {
  const factory LinkedProducerAccount({
    @JsonKey(name: 'producer_account_id') required String producerAccountId,
    required String name,
  }) = _LinkedProducerAccount;

  factory LinkedProducerAccount.fromJson(Map<String, Object?> json) =>
      _$LinkedProducerAccountFromJson(json);
}

File: back/persistence/model/src/main/kotlin/ProducerAccount.kt
@file:OptIn(ExperimentalTime::class)

package persistence.model

import id.Id
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.time.ExperimentalTime
import kotlin.time.Instant

/**
 * Producer entity that groups members working for a production unit with embedded users, organizations and products.
 */
@Serializable
data class ProducerAccount(
    @SerialName("producer_account_id") val producerAccountId: Id<ProducerAccount>,
    val name: String,
    @SerialName("contact_email") val contactEmail: String? = null,
    val address: String? = null,
    val website: String? = null,
    @SerialName("active_status") val activeStatus: Boolean,
    @SerialName("created_instant") val createdInstant: Instant,
    @SerialName("last_updated_instant") val lastUpdatedInstant: Instant,
    val users: List<ProducerUser> = emptyList(),
    val organizations: List<ProducerOrganization> = emptyList(),
    val products: List<ProducerProduct> = emptyList(),
    @SerialName("user_preferences") val userPreferences: UserPreferences =
        UserPreferences(
            emailNotificationsEnabled = true,
            pushNotificationsEnabled = false,
            lastUpdatedInstant = Instant.fromEpochMilliseconds(0L),
        ),
    @SerialName("management_mode") val managementMode: ProducerManagementMode = ProducerManagementMode.ACCOUNT_BACKED,
    @SerialName("linked_producer_account") val linkedProducerAccount: LinkedProducerAccount? = null,
)

@Serializable
enum class ProducerManagementMode {
    ACCOUNT_BACKED,
    NO_ACCOUNT,
}

@Serializable
data class LinkedProducerAccount(
    @SerialName("producer_account_id") val producerAccountId: Id<ProducerAccount>,
    val name: String,
)

/**
 * Producer user association (embedded in ProducerAccount).
 */
@Serializable
data class ProducerUser(
    @SerialName("producer_id") val producerId: Id<Producer>,
    val role: ProducerRole,
    @SerialName("association_instant") val associationInstant: Instant,
    val status: ProducerStatus,
)

/**
 * Producer-Organization association (embedded in ProducerAccount).
 */
@Serializable
data class ProducerOrganization(
    @SerialName("organization_id") val organizationId: Id<Organization>,
    @SerialName("association_instant") val associationInstant: Instant,
    val status: OrganizationProducerStatus,
)

/**
 * Product definition for this producer (embedded in ProducerAccount).
 */
@Serializable
data class ProducerProduct(
    val name: String,
    @SerialName("product_type_id") val productTypeId: Id<ProductType>,
    @SerialName("supported_basket_sizes") val supportedBasketSizes: List<BasketSize>,
    val description: String? = null,
)
